source/backtraceMetric.ts

```diff
@@ -0,0 +1,226 @@
+      this.sendSummedEvent(newSessionId, 'application-launches');
+
+      // If sessionId is set and lastActive is over 30 minutes ago, create new session only.
+    } else if (this.lastActive < this.CURRENT_TIMESTAMP - this.THIRTY_MINUTES) {
+      this.createNewSession();
+      this.sendUniqueEvent(this.sessionId);
+      // can also send "session length" by sending lastActive - sessionStart.
+      // This can get a sense of "session length" / "time spent on app", "session free minutes", "avg time per error" etc
+    }
+
+    // handle summed events
+    // ex. page load / event
+  }
+
+  /**
+   * Send POST to unique-events API endpoint
+   */
+  private async sendUniqueEvent(sessionId: string): Promise<void> {
+    const endpoint = `https://${this.DIRECTORY}.backtrace.io/api/unique-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    const payload = {
+      application: this.applicationName,
+      appversion: __VERSION__,
+      metadata: {
+        dropped_events: 0,
+      },
+      unique_events: [
+        {
+          timestamp: this.CURRENT_TIMESTAMP,
+          unique: sessionId,
+          attributes: {
+            guid: sessionId,
+            'uname.sysname': this.userAgent,
+            'application.version': __VERSION__,
+            'application.session': '', // application session ?
+          },
+        },
+      ],
+    };
+
+    await this.post(endpoint, JSON.stringify(payload));
+    return;
+  }
+
+  /**
+   * Send POST to summed-events API endpoint
+   */
+  private async sendSummedEvent(
+    sessionId: string,
+    metricGroup: string,
+  ): Promise<void> {
+    const endpoint = `https://${this.DIRECTORY}.backtrace.io/api/summed-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    const payload = {
+      application: this.applicationName,
+      appversion: __VERSION__,
+      metadata: {
+        dropped_events: 0,
+      },
+      summed_events: [
+        {
+          timestamp: this.CURRENT_TIMESTAMP,
+          metric_group: metricGroup,
+          attributes: {
+            guid: sessionId,
+            'uname.sysname': this.userAgent,
+            'application.version': __VERSION__,
+            'application.session': '', // application session ?
+          },
+        },
+      ],
+    };
+
+    await this.post(endpoint, JSON.stringify(payload));
+    return;
+  }
+
+  /**
+   * Send POST request.
+   * @param url - string endpoint
+   * @param data - stringified JSON object
+   */
+  private async post(url: string, data: string): Promise<void> {
+    try {
+      return new Promise<void>((res, rej) => {
+        const http = new XMLHttpRequest();
+        http.setRequestHeader('Content-type', 'application/json');
+        http.timeout = this.timeout;
+        http.open('POST', url, true);
+        http.send(data);
+        http.onload = (e) => {
+          if (http.readyState === XMLHttpRequest.DONE) {
```

Comment:
small nit:
You can save one indent just by checking if the readyState is not equal to ` XMLHttpRequest.DONE` . With one less indent, your code looks smaller/the line of code isn't too large.


```suggestion
          if (http.readyState !== XMLHttpRequest.DONE) {
              return;
          }
```

Reply:
Nice! updated